### DISCUSSION TITLE
Exploding the array of JSON objects

### DISCUSSION THREAD

**Discussion by omurov** [vectordotdev/vector#10391] [Q&A] [answered]
How can we explode/unnest the array of JSON objects, for example:

{ "message": "[ { \"file\": \"/var/log/nginx/API.access.log\", \"host\": \"platform-testing-api2\", \"json\": { \"agent\": \"Go-http-client/1.1\", \"body_sent\": { \"bytes\": 35 }, \"http_version\": \"HTTP/1.1\", \"method\": \"GET\", \"referrer\": \"\", \"remote_ip\": \"172.17.0.3\", \"request_time\": 0.052, \"response_code\": 200, \"time\": \"2021-12-10T15:27:09+02:00\", \"upstream_cache_status\": \"MISS\", \"url\": \"/api/health/dino\", \"user_name\": \"\" }, \"source_type\": \"file\", \"timestamp\": \"2021-12-10T13:27:10.456924207Z\" }, { \"file\": \"/var/log/nginx/API.access.log\", \"host\": \"platform-testing-api2\", \"json\": { \"agent\": \"\", \"body_sent\": { \"bytes\": 0 }, \"http_version\": \"HTTP/1.0\", \"method\": \"GET\", \"referrer\": \"\", \"remote_ip\": \"172.18.201.21\", \"request_time\": 0.015, \"response_code\": 200, \"time\": \"2021-12-10T15:27:10+02:00\", \"upstream_cache_status\": \"\", \"url\": \"/api/health\", \"user_name\": \"\" }, \"source_type\": \"file\", \"timestamp\": \"2021-12-10T13:27:10.975833700Z\" } ]" }

When we use:
. = parse_json!(.message) 

we get the following:
[{ "file": "/var/log/nginx/API.access.log", "host": "platform-testing-api2", "json": { "agent": "Go-http-client/1.1", "body_sent": { "bytes": 35 }, "http_version": "HTTP/1.1", "method": "GET", "referrer": "", "remote_ip": "172.17.0.3", "request_time": 0.052, "response_code": 200, "time": "2021-12-10T15:27:09+02:00", "upstream_cache_status": "MISS", "url": "/api/health/dino", "user_name": "" }, "source_type": "file", "timestamp": "2021-12-10T13:27:10.456924207Z" }, { "file": "/var/log/nginx/API.access.log", "host": "platform-testing-api2", "json": { "agent": "", "body_sent": { "bytes": 0 }, "http_version": "HTTP/1.0", "method": "GET", "referrer": "", "remote_ip": "172.18.201.21", "request_time": 0.015, "response_code": 200, "time": "2021-12-10T15:27:10+02:00", "upstream_cache_status": "", "url": "/api/health", "user_name": "" }, "source_type": "file", "timestamp": "2021-12-10T13:27:10.975833700Z" }]

How can we explode this dynamically? as we can get 1..N JSON messages inside an array.

**Comment by spencergilbert** [comment, marked as answer]
`. = parse_json!(.message)` and getting an array of objects should pass each object as a separate event to the next component, is that not the behavior you're seeing?

There is also an [`unnest`](https://vector.dev/docs/reference/vrl/functions/#unnest) function.

**Comment by omurov** [comment]
Hello Spencer,

Thank you for your answer, we have tried those, but it does not split this
into separate events. If I test run it and output to the file it writes the
whole array to one line.
For unnset there is no path that I can provide, as it’s just an array of
json objects, unnset asks for the path to explode based on that.

Regards,
Alex

**Reply by spencergilbert** [reply]
Given this log and config:

```shell
❯ cat /tmp/explode.log
[ { "file": "/var/log/nginx/API.access.log", "host": "platform-testing-api2", "json": { "agent": "Go-http-client/1.1", "body_sent": { "bytes": 35 }, "http_version": "HTTP/1.1", "method": "GET", "referrer": "", "remote_ip": "172.17.0.3", "request_time": 0.052, "response_code": 200, "time": "2021-12-10T15:27:09+02:00", "upstream_cache_status": "MISS", "url": "/api/health/dino", "user_name": "" }, "source_type": "file", "timestamp": "2021-12-10T13:27:10.456924207Z" }, { "file": "/var/log/nginx/API.access.log", "host": "platform-testing-api2", "json": { "agent": "", "body_sent": { "bytes": 0 }, "http_version": "HTTP/1.0", "method": "GET", "referrer": "", "remote_ip": "172.18.201.21", "request_time": 0.015, "response_code": 200, "time": "2021-12-10T15:27:10+02:00", "upstream_cache_status": "", "url": "/api/health", "user_name": "" }, "source_type": "file", "timestamp": "2021-12-10T13:27:10.975833700Z" } ]

❯ cat /tmp/explode.yaml
api:
  enabled: true
sources:
  file:
    type: file
    include: ["/tmp/explode.log"]
    ignore_checkpoints: true
transforms:
  remap:
    type: remap
    inputs: ["file"]
    source: |
      . = parse_json!(.message)
sinks:
  console:
    type: console
    inputs: ["remap"]
    encoding:
      codec: json
```

I get two separate events written to the console sink:

```json
{"file":"/var/log/nginx/API.access.log","host":"platform-testing-api2","json":{"agent":"Go-http-client/1.1","body_sent":{"bytes":35},"http_version":"HTTP/1.1","method":"GET","referrer":"","remote_ip":"172.17.0.3","request_time":0.052,"response_code":200,"time":"2021-12-10T15:27:09+02:00","upstream_cache_status":"MISS","url":"/api/health/dino","user_name":""},"source_type":"file","timestamp":"2021-12-10T13:27:10.456924207Z"}
{"file":"/var/log/nginx/API.access.log","host":"platform-testing-api2","json":{"agent":"","body_sent":{"bytes":0},"http_version":"HTTP/1.0","method":"GET","referrer":"","remote_ip":"172.18.201.21","request_time":0.015,"response_code":200,"time":"2021-12-10T15:27:10+02:00","upstream_cache_status":"","url":"/api/health","user_name":""},"source_type":"file","timestamp":"2021-12-10T13:27:10.975833700Z"}
```

**Comment by omurov** [comment]
Thank you, will try it again. For some reason writing to file does not
separate those lines.

**Reply by spencergilbert** [reply]
Interesting, I replaced the console out with:

```yaml
sinks:
  file_out:
    type: file
    inputs: ["remap"]
    path: /tmp/vector-%Y-%m-%d.log
    encoding:
      codec: ndjson
```

And still get the same behavior: 

```json
❯ cat /tmp/vector-2021-12-10.log
{"file":"/var/log/nginx/API.access.log","host":"platform-testing-api2","json":{"agent":"Go-http-client/1.1","body_sent":{"bytes":35},"http_version":"HTTP/1.1","method":"GET","referrer":"","remote_ip":"172.17.0.3","request_time":0.052,"response_code":200,"time":"2021-12-10T15:27:09+02:00","upstream_cache_status":"MISS","url":"/api/health/dino","user_name":""},"source_type":"file","timestamp":"2021-12-10T13:27:10.456924207Z"}
{"file":"/var/log/nginx/API.access.log","host":"platform-testing-api2","json":{"agent":"","body_sent":{"bytes":0},"http_version":"HTTP/1.0","method":"GET","referrer":"","remote_ip":"172.18.201.21","request_time":0.015,"response_code":200,"time":"2021-12-10T15:27:10+02:00","upstream_cache_status":"","url":"/api/health","user_name":""},"source_type":"file","timestamp":"2021-12-10T13:27:10.975833700Z"}
```

Definitely interested to see what's happening. Please share a complete config (if you can) if the issue continues.
